compiler/rustc_trait_selection/src/traits/error_reporting/on_unimplemented.rs

```diff
@@ -824,7 +825,11 @@ impl<'tcx> OnUnimplementedFormatString {
             .filter_map(|param| {
                 let value = match param.kind {
                     GenericParamDefKind::Type { .. } | GenericParamDefKind::Const { .. } => {
-                        trait_ref.args[param.index as usize].to_string()
+                        if let Some(ty) = trait_ref.args[param.index as usize].as_type() {
+                            tcx.short_ty_string(ty, &mut None)
```

Comment:
I think you should modify `format` to take an extra `file: &mut None` parameter to pass in here and modify `OnUnimplementedNote` to carry `file` as well. You can see how we present that [in other places](https://github.com/rust-lang/rust/blob/86a7fc840f11cd1f20fdf26d3071e34a2d4bc313/compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs#L1272-L1289), which makes it so that if the user is interested in looking at the full type for whatever reason, they can. Ideally we'd have a single `file` for the entire error, and pass that around everywhere that we use `short_ty_string` for E0277, particularly `note_obligation_cause_code`, but that doesn't *need* to be addressed in this PR, but since you're adding another use there it might be a good idea to do so. Then we can remove all of the redundant "the full type name has been written to '{}'" notes and push them to the top-most call of `note_obligation_cause_code`.

Reply:
Not-within-the-scope-of-this-PR remark: the long type names when written to disk is... still incredibly long. I previously saw some issue proposing that the long types written to disk should be pretty printed (but now I cannot find it) and I think that might be good as an option (not necessarily default-behavior because it's probably easier to diff two long types when they are in single lines).